### PR TITLE
RasaModelData can handle 4D Tensors

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
- RasaModelData can handle 4D Tensors

related to https://github.com/RasaHQ/rasa/issues/6670

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tabergma** [comment]
We have a lot of test failures. But I guess most of them or all of them come from the `e2e` itself. Should we fix them first on `e2e` branch itself?

**Comment by Ghostvv** [comment]
> Should we fix them first on e2e branch itself?
I didn't look into tests there 🙈

File: rasa/utils/tensorflow/model_data.py [hunk 1]
```diff
@@ -458,21 +514,37 @@ def _get_shapes_types(self) -> Tuple:
...
-            if isinstance(features[0], scipy.sparse.spmatrix):
-                # scipy matrix is converted into indices, data, shape
-                shapes.append((None, features[0].ndim + 1))
-                shapes.append((None,))
-                shapes.append((features[0].ndim + 1))
-            elif features[0].ndim == 0:
-                shapes.append((None,))
-            elif features[0].ndim == 1:
-                shapes.append((None, features[0].shape[-1]))
+        def append_shape(_features: Union[Features_4D, Features_3D]) -> None:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:517]: it looks like in the rest of the code, `features` correspond to `Union[List[Features_4D], List[Features_3D]]` shall we create type for it?
and rename `_features` here to smth more explicit?

File: rasa/utils/tensorflow/model_data.py [hunk 2]
```diff
@@ -31,14 +31,22 @@
...
 #   "numpy array containing sparse features for every training example"
 # ]}
-Data = Dict[Text, Dict[Text, List[np.ndarray]]]
+# In case of e2e data, we also deal with 4d tensors, e.g.
+# "action_text" -> { "sequence": [
+#   [["numpy array containing dense features for every training example"]],
+#   [["numpy array containing sparse features for every training example"]]
+# ]}
+Features_4D = List[np.ndarray]
+Features_3D = np.ndarray
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:40]: hm, so list in 4D correspond to batch dim?, however in 3D it's first dim of ndarray.
I'm afraid it's prone to confusion after some time. Shall we create a special structure for Features, that would be basically ndarray, but contain all the info: whether it is sparse, whether it is 4D, which dimensions are what etc?

File: rasa/utils/tensorflow/model_data.py [hunk 3]
```diff
@@ -323,14 +361,18 @@ def split(
...
+                if RasaModelData.is_in_4d_format(f)
+                else f[counts > 1]
                 for attribute_data in self.data.values()
                 for features in attribute_data.values()
                 for f in features
             ]
             # collect data points that are unique for their label
             solo_values = [
-                f[counts == 1]
+                list(np.array(f)[counts == 1])
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:373]: that is a very strange thing 😂, I'd say it'd be better to use some special structure, to avoid things like that

File: rasa/utils/tensorflow/model_data.py (hunk 2 above)
Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: and it seems to be if we keep the outer `ndarray`, it'll simplify things in the code. Can we subclass `ndarray`?

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:40]: why would you subclass `ndarray`?

I was first thinking of keeping both as numpy array with different dimensionalities, but then you would need to pad the data event before adding it to `RasaModelData` when creating the 4D features, which would have complicated the model data creation in TED. So, what do you suggest as a structure?

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: but the ndarray in 3D features also contain non padded data, basically serving as a list

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: > why would you subclass ndarray? what do you suggest as a structure?

smth that maintain the slicing of ndarray (needed for operating on batch dimension), but also containing the info about sparse, 4D, etc. to simplify our checks

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: basically we're already using `ndarray` as ragged array. Because we're padding it only at last moment before creating tf.tensors

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: I think we can even use ragged tensors, avoiding need for manual padding https://www.tensorflow.org/guide/ragged_tensor

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:40]: So, you are saying 4D should become np.array(np.array(np.array)) and 3D np.array(np.array)? And then we wrap it into some kind of structure so that we know if it is sparse, 4D, etc. So basically this object would already contain the information around FeatureSignature. Can try that.

I guess ragged tensors does not work for sparse, does it?

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:40]: ok, will try something, will ping you once it is ready for review

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: I meant ragged tensors for dense

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:40]: >  3D np.array(np.array)

I believe we already have it

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:40]: yes, we do, just wanted to clarify

**Comment by evgeniiaraz** [comment]
@tabergma I think I'll slack off the review here because I'd like to take my time to understand everything but I feel like by the time I do, I might not have proper access to Rasa github 😅 (not sure what the policies here are)

**Comment by Ghostvv** [comment]
@evgeniiaraz this PR is public PR 😜

File: rasa/utils/tensorflow/model_data.py [hunk 4]
```diff
@@ -24,21 +24,85 @@
...
+    ):
+        feature_array = np.asarray(input_array).view(cls)
+
+        if number_of_dimensions <= 2:
+            feature_array.feature_dimension = input_array.shape[-1]
+            feature_array.is_sparse = isinstance(input_array[0], scipy.sparse.spmatrix)
+        elif number_of_dimensions == 3:
+            feature_array.feature_dimension = input_array[0].shape[-1]
+            feature_array.is_sparse = isinstance(input_array[0], scipy.sparse.spmatrix)
+        else:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:41]: should we explicitly put `elif number_of_dimensions == 4:` and `else: raise`

File: rasa/utils/tensorflow/model_data.py [hunk 5]
```diff
@@ -96,7 +162,7 @@ def items(self) -> ItemsView:
         """
         return self.data.items()
 
-    def values(self) -> ValuesView[Dict[Text, List[np.ndarray]]]:
+    def values(self,) -> ValuesView[Dict[Text, List[FeatureArray]]]:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:165]:
```suggestion
    def values(self) -> ValuesView[Dict[Text, List[FeatureArray]]]:
```

File: rasa/utils/tensorflow/model_data.py [hunk 6]
```diff
@@ -209,9 +275,9 @@ def feature_dimension(self, key: Text, sub_key: Text) -> int:
             return 0
 
         number_of_features = 0
-        for data in self.data[key][sub_key]:
-            if data.size > 0:
-                number_of_features += data[0].shape[-1]
+        for features in self.data[key][sub_key]:
+            if len(features) > 0:
+                number_of_features += features.feature_dimension
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:280]: let's rename `number_of_features`

File: rasa/utils/tensorflow/model_data.py [hunk 7]
```diff
@@ -24,21 +24,85 @@
...
+        if obj is None:
+            return
+
+        self.feature_dimension = getattr(obj, "feature_dimension", None)
+        self.number_of_dimensions = getattr(obj, "number_of_dimensions", None)
+        self.is_sparse = getattr(obj, "is_sparse", None)
+
+        default_attributes = {
+            "feature_dimension": self.feature_dimension,
+            "number_of_dimensions": self.number_of_dimensions,
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:61]: I think `feature_dimension` is a confusing name (since features have multiple dimensions). Not sure what would be better name. In keras this number is called `units`.

File: rasa/utils/tensorflow/model_data.py [hunk 8]
```diff
@@ -458,21 +531,35 @@ def _get_shapes_types(self) -> Tuple:
...
-                # scipy matrix is converted into indices, data, shape
-                shapes.append((None, features[0].ndim + 1))
-                shapes.append((None,))
-                shapes.append((features[0].ndim + 1))
-            elif features[0].ndim == 0:
-                shapes.append((None,))
-            elif features[0].ndim == 1:
-                shapes.append((None, features[0].shape[-1]))
+        def append_shape(_features: FeatureArray) -> None:
+            if _features.number_of_dimensions == 4:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:535]: is there a reason why `number_of_dimensions == 4` is first? Can it be put as one of the elifs below?

File: rasa/utils/tensorflow/model_data.py [hunk 9]
```diff
@@ -458,21 +531,35 @@ def _get_shapes_types(self) -> Tuple:
...
-            if isinstance(features[0], scipy.sparse.spmatrix):
-                # scipy matrix is converted into indices, data, shape
-                shapes.append((None, features[0].ndim + 1))
-                shapes.append((None,))
-                shapes.append((features[0].ndim + 1))
-            elif features[0].ndim == 0:
-                shapes.append((None,))
-            elif features[0].ndim == 1:
-                shapes.append((None, features[0].shape[-1]))
+        def append_shape(_features: FeatureArray) -> None:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:534]: can this method become a part of `FeatureArray` class?

File: rasa/utils/tensorflow/model_data.py [hunk 10]
```diff
@@ -843,7 +939,39 @@ def _pad_dense_data(array_of_dense: np.ndarray) -> np.ndarray:
         return data_padded.astype(np.float32)
 
     @staticmethod
-    def _scipy_matrix_to_values(array_of_sparse: np.ndarray) -> List[np.ndarray]:
+    def _pad_4d_dense_data(array_of_array_of_dense: FeatureArray) -> np.ndarray:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:942]: ragged tensors are not suitable for us?

File: rasa/utils/tensorflow/models.py [hunk 11]
```diff
@@ -285,7 +285,7 @@ def train_on_batch(
         self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
 
     def build_for_predict(
-        self, predict_data: RasaModelData, eager: bool = False
+        self, predict_data: RasaModelData, eager: bool = True
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/models.py:288]:
```suggestion
        self, predict_data: RasaModelData, eager: bool = False
```

File: rasa/utils/tensorflow/models.py [hunk 12]
```diff
@@ -541,6 +541,22 @@ def batch_to_model_data_format(
...
+                        batch_data[key][sub_key].append(
+                            tf.SparseTensor(
+                                batch[idx],
+                                batch[idx + 1],
+                                [
+                                    batch[idx + 2][0],
+                                    batch[idx + 2][1],
+                                    batch[idx + 2][2],
+                                    feature_dimension,
+                                ],
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/models.py:556]: isn't it easier: `shape = [batch[idx + 2][i] for i in range(number_of_dimensions - 1)] + [feature_dimension]`? then we don't need the `if` about `number_of_dimensions`

**Review by Ghostvv — COMMENTED** [review]
looks great! I think it is way more convenient now, thanks for trying it

File: rasa/utils/tensorflow/model_data.py (hunk 10 above)
Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:942]: You are too fast with reviewing :smile:  It is not yet fully working, wanted to look into that once everything else works. Please be patient :slightly_smiling_face:

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py:942]: oh sorry 😅 was excited to look how subclassing ndarray idea worked out

**Comment by tabergma** [comment]
Implementation of `FeatureArray` is done. Did not had time to take a look at ragged tensors. Will do that on Monday.

File: rasa/utils/tensorflow/model_data.py (hunk 10 above)
Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:1000]: @Ghostvv `RaggedTensors` look good, however they are currently not supported by `tf.data.Dataset.from_generator`... It is already fixed (https://github.com/tensorflow/tensorflow/pull/41981), but seem to be not yet released. We would also need to update some of our layers I guess. I'll create an issue for it to update it in the future.

Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:1000]: https://github.com/RasaHQ/rasa/issues/6909

File: rasa/utils/tensorflow/model_data.py [hunk 13]
```diff
@@ -152,7 +157,7 @@ def _validate_number_of_dimensions(
                 f"array: {input_array}."
             )
 
-    def get_shape_type_info(self) -> Tuple[List[Tuple], List[Tuple]]:
+    def get_shape_type_info(self) -> Tuple[List, List]:
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data.py:160]: do we know what these `List`s contain?

File: rasa/utils/tensorflow/model_data_utils.py [hunk 14]
```diff
@@ -186,8 +186,8 @@ def _features_for_attribute(
         tracker_features, zero_state_features[attribute]
     )
 
-    sparse_features = defaultdict(list)
-    dense_features = defaultdict(list)
+    sparse_features = defaultdict()
+    dense_features = defaultdict()
```
Comment (Ghostvv) [inline rasa/utils/tensorflow/model_data_utils.py:190]: do we need a `defaultdict`? can't it simply be a `dict`?

File: rasa/utils/tensorflow/model_data.py (hunk 13 above)
Comment (tabergma) [reply rasa/utils/tensorflow/model_data.py:160]: It depends on the array. Sometimes it is a list of tuples of ints, sometimes it is just an int. I could change it to
```
Tuple[List[Union[int, Tuple[None], Tuple[None, int], Tuple[None, None, int], Tuple[None, None, None, int]]], List[int]]
```

Comment (Ghostvv) [reply rasa/utils/tensorflow/model_data.py]: oh no

**Review by Ghostvv — APPROVED** [review]
Looks great! one small thing maybe: sparse 4D and 3D methods, seems to have quite a bit of shared code. Does it make sense to unite them somehow?

**Comment by tabergma** [comment]
They actually do not really share that much code. The logic is the same but for 4D we always have an additional index or for loop. So if we would to extract some of the logic you would have a lot of if else statements I guess. So might be not worth it. 

The tests that are failing should be related to the changes in `e2e`. I guess we just merge and solve them later?

**Comment by Ghostvv** [comment]
yes, let's merge then
